---
title: "Tools"
# layout: categories
permalink: /tools/
layout: single_center
author_profile: false
classes: wide
#collection: tutorial
---

This page is intended for me to keep records of scripts that I've written.

To be updated.

<br>

> **Convert gbk (GenBank files) to csv:**
- Tutorial: [gbk to csv](/tutorials/gbk_to_csv)
- Python script: [extract_gbk_to_csv.py](https://github.com/NanamiKubota/NanamiKubota.github.io/blob/main/scripts/extract_gbk_to_csv.py)

<br>

> **Convert breseq output to csv using gdtools:**
- Tutorial: [breseq parser gdtools](/tutorials/breseq_parser_gdtools)
- Python script: [breseq_parser_gdtools.py](https://github.com/NanamiKubota/NanamiKubota.github.io/blob/main/scripts/breseq_parser_gdtools.py)
- Example shell script to run python script on beagle: [breseq_parser_gdtools.sh](https://github.com/NanamiKubota/NanamiKubota.github.io/blob/main/scripts/breseq_parser_gdtools.sh)

<br>

> **Compare amino acid sequence of homologous loci (i.e. multiple sequence align; MSA) across several Genbank files and output summary of differences as a csv:**
- Under construction
  
<!-- - Tutorial: [Multiple Sequence Align (MSA) Genbank / Reference Sequences](/tutorials/msa_gb_ref_seq)
- Python script: [msa_comparison.py](https://github.com/NanamiKubota/NanamiKubota.github.io/blob/main/scripts/msa_comparison.py) -->

<br>

Work in Progress:

> **Convert Genbank (gbk) files to Roary-acceptable gff3 file:**
- Coming soon. You can use this tool to convert a gbk file to Roary-acceptable gff3 file without having to run Prokka.